### PR TITLE
Add Git repo awareness to GPG/Pass credential store

### LINKED ISSUE #657: Using GPG/Pass for credential storage overwrites even without changes
**Which version of GCM are you using?**

Git Credential Manager version `2.0.632+478753bbb7` (Linux)

**Expected behavior**

When using `git config --global credential.credentialStore gpg` with GCM, files under `git/https/<service>/<username>.gpg` should be updated only when the credentials change 

**Actual behavior**

The git/https/github.com/<username>.gpg file is updated on every interaction with a remote repository, such as pull/push

**Logs**

[log.txt](https://github.com/GitCredentialManager/git-credential-manager/files/8446144/log.txt)

### PR DESCRIPTION
**Pull request by mjcheetham**
Update the GPG/Pass credential store to be aware of stores that are also Git repositories.

Like the `pass` command-line tool, if and when a credential file is edited (add/update/delete) and the store has been initialised as a Git repository, GCM will now add the file change and create a commit.

We make sure we don't fail the store/remove operation if the Git action fails - this is a best effort.

Should also help fix #657 along side with #739

### CODE REVIEW

File: src/shared/Core/Git.cs [hunk 1]
```diff
@@ -219,6 +225,35 @@ public void AddFile(string filePath)
             }
         }
 
+        public void Commit(string message)
```
Comment (ldennington) [inline src/shared/Core/Git.cs:228]: Is there an opportunity to share some of this code with the `Add` method? May not be worthwhile, but was just a thought I had.

File: src/shared/Core/Git.cs [hunk 2]
```diff
@@ -219,6 +225,35 @@ public void AddFile(string filePath)
...
+        {
+            EnsureArgument.NotNullOrEmpty(message, nameof(message));
+
+            using (var git = CreateProcess("commit -F -"))
```
Comment (ldennington) [inline src/shared/Core/Git.cs:232]: Why are we using `-F` here? Isn't that for [reading a commit message from a file](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--Fltfilegt)?

File: src/shared/Core.Tests/Interop/Posix/GnuPassCredentialStoreTests.cs [hunk 3]
```diff
@@ -89,7 +90,105 @@ public void GnuPassCredentialStore_Remove_NotFound_ReturnsFalse()
...
+
+                mockGit.Verify(x => x.AddFile(expectedFilePath), Times.Once);
+                mockGit.Verify(x => x.Commit(It.IsAny<string>()), Times.Once);
+                mockGit.Reset();
+
+                // Read
+                ICredential outCredential = collection.Get(service, userName);
+
+                Assert.NotNull(outCredential);
+                Assert.Equal(userName, userName);
```
Comment (ldennington) [inline src/shared/Core.Tests/Interop/Posix/GnuPassCredentialStoreTests.cs:135]: Should we be comparing to `outcredential.Account`?

File: src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs [hunk 4]
```diff
@@ -90,6 +111,86 @@ protected override void SerializeCredential(FileCredential credential)
...
+                sb.AppendFormat("[GCM] Update credential {0}", credential.Service);
+                if (hasAccount)
+                {
+                    sb.AppendFormat(" ({0})", credential.Account);
+                }
+
+                sb.AppendLine();
+                sb.AppendLine();
+                sb.AppendFormat("Update Git Credential Manager credential:{0}", Environment.NewLine);
+                sb.AppendFormat("  Service: {1}{0}", Environment.NewLine, credential.Service);
```
Comment (ldennington) [inline src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs:130]: Should these be reversed (as they are below)?

```suggestion
                sb.AppendFormat("  Service: {1}{0}", credential.Service, Environment.NewLine);
```

File: src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs [hunk 5]
```diff
@@ -90,6 +111,86 @@ protected override void SerializeCredential(FileCredential credential)
...
+                if (hasAccount)
+                {
+                    sb.AppendFormat("  Account: {1}{0}", Environment.NewLine, credential.Account);
```
Comment (ldennington) [inline src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs:133]: Should these be reversed (as they are below)?
```suggestion
                    sb.AppendFormat("  Account: {1}{0}", credential.Account, Environment.NewLine);
```

File: src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs [hunk 6]
```diff
@@ -90,6 +111,86 @@ protected override void SerializeCredential(FileCredential credential)
...
+                catch (GitException ex)
+                {
+                    // Don't fail just because we failed to commit in the pass store Git repo
+                    Trace.WriteLine("Failed to update Git repository in pass store");
+                    Trace.WriteException(ex);
+                }
+            }
+        }
+
+        protected override bool DestroyCredential(FileCredential credential)
```
Comment (ldennington) [inline src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs:153]: Potentially consider extracting some of the shared logic between this and `SerializeCredential`.

**Review by ldennington — COMMENTED** [review]
Thanks for the nice, clean commit series. Made this a pleasure to look through.

File: src/shared/Core/Git.cs (hunk 2 above)
Comment (IjobaFifti) [reply src/shared/Core/Git.cs:232]: Good

File: src/shared/Core.Tests/Interop/Posix/GnuPassCredentialStoreTests.cs (hunk 3 above)
Comment (IjobaFifti) [reply src/shared/Core.Tests/Interop/Posix/GnuPassCredentialStoreTests.cs:135]: Good

File: src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs (hunk 4 above)
Comment (IjobaFifti) [reply src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs:130]: Good

File: src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs (hunk 6 above)
Comment (IjobaFifti) [reply src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs:153]: G

**Comment by IjobaFifti** [comment]
I love to know more about Cpypto currencies

File: src/shared/Core/Git.cs (hunk 2 above)
Comment (IjobaFifti) [reply src/shared/Core/Git.cs:232]: Love For It

File: src/shared/Core/Git.cs (hunk 1 above)
Comment (IjobaFifti) [reply src/shared/Core/Git.cs:228]: I can understand

File: src/shared/Core.Tests/GitTests.cs [hunk 7]
```diff
@@ -203,6 +203,55 @@ public void Git_GitPath_ReturnsGitPath()
...
+            File.WriteAllText(filePath, "test file contents");
+
+            string gitPath = GetGitPath();
+            var trace = new NullTrace();
+            var env = new TestEnvironment();
+
+            var git = new GitProcess(trace, env, gitPath, workDirPath);
+            git.AddFile(filePath);
+
+            GitResult lsResult = ExecGit(repoPath, workDirPath, "ls-files --cached");
```
[the remark this answers is not in the document]
Comment (IjobaFifti) [reply src/shared/Core.Tests/GitTests.cs:220]: Can I learn more?

**Review by mrarif6585 — COMMENTED** [review]
mjcheetham:gpg-git

**Review by mrarif6585 — COMMENTED** [review]
src/shared/Core.Tests/GitTests.cs

**Comment by ebibesty** [comment]
Please I don't know you.

**Review by AdrianaDJ — COMMENTED** [review]
Create dependabot.yml deployment and create a new application

File: src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs (hunk 6 above)
Comment (16CentAstrology) [reply src/shared/Core/Interop/Posix/GpgPassCredentialStore.cs:153]: > Potentially consider extracting some of the shared logic between this and `Serialize Credential`.

**Review by Bombay33266 — COMMENTED** [review]
Monroe

**Comment by mjcheetham** [comment]
Closing as this should no longer be an issue now that we only write credentials to the store when they have been modified.

**Pull request closed without merging**